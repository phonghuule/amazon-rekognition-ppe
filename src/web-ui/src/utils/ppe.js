const capitalize = (s) => s.charAt(0).toUpperCase() + s.slice(1).toLowerCase();
const percentageToString = (percentage) => Math.floor(percentage * 10) / 10;
const formatText = (t) => t.replace(/_/gi, " ").toLowerCase();

export const ppeMapper = (person) => {
  const bodyParts = (person.BodyParts || []).filter(
    (x) => x.EquipmentDetections && x.EquipmentDetections.length > 0
  );

  const results = bodyParts
    .map((p) =>
      p.EquipmentDetections.map((eq) => ({
        bodyPart: formatText(p.Name),
        confidence: percentageToString(p.Confidence),
        type: capitalize(formatText(eq.Type)),
        coversBodyPart: eq.CoversBodyPart.Value,
        coversBodyPartConfidence: percentageToString(
          eq.CoversBodyPart.Confidence
        ),
        boundingBox: eq.BoundingBox,
      }))
    )
    .flat();

  return {
    id: person.Id,
    boundingBox: person.BoundingBox,
    results,
  };
};
