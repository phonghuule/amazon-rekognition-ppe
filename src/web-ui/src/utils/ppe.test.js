import { ppeMapper } from "./ppe";

const testData = {
  Persons: [
    {
      BodyParts: [
        {
          Confidence: 99.99930572509766,
          EquipmentDetections: [
            {
              BoundingBox: {
                Height: 0.3095385730266571,
                Left: 0.2645637094974518,
                Top: 0.29297786951065063,
                Width: 0.1740061491727829,
              },
              Confidence: 98.30194854736328,
              CoversBodyPart: {
                Confidence: 79.82731628417969,
                Value: true,
              },
              Type: "MASK",
            },
            {
              BoundingBox: {
                Height: 0.3095385730266571,
                Left: 0.2645637094974518,
                Top: 0.29297786951065063,
                Width: 0.1740061491727829,
              },
              Confidence: 98.30194854736328,
              CoversBodyPart: {
                Confidence: 79.82731628417969,
                Value: true,
              },
              Type: "FOO",
            },
          ],
          Name: "FACE",
        },
        {
          Confidence: 99.99846649169922,
          EquipmentDetections: [
            {
              BoundingBox: {
                Height: 0.37994688749313354,
                Left: 0.342035174369812,
                Top: 0.16647937893867493,
                Width: 0.2135859727859497,
              },
              Confidence: 85.8783187866211,
              CoversBodyPart: {
                Confidence: 84.20125579833984,
                Value: true,
              },
              Type: "HELMET",
            },
          ],
          Name: "HEAD",
        },
        {
          Confidence: 99.72054290771484,
          EquipmentDetections: [
            {
              BoundingBox: {
                Height: 0.33101364970207214,
                Left: 0.5892353653907776,
                Top: 0.23174917697906494,
                Width: 0.18587328493595123,
              },
              Confidence: 96.79559326171875,
              CoversBodyPart: {
                Confidence: 100,
                Value: true,
              },
              Type: "GLOVE",
            },
          ],
          Name: "LEFT_HAND",
        },
        {
          Confidence: 99.87159729003906,
          EquipmentDetections: [
            {
              BoundingBox: {
                Height: 0.32489436864852905,
                Left: 0.20085042715072632,
                Top: 0.2759649455547333,
                Width: 0.18243715167045593,
              },
              Confidence: 78.28016662597656,
              CoversBodyPart: {
                Confidence: 100,
                Value: true,
              },
              Type: "GLOVE",
            },
          ],
          Name: "RIGHT_HAND",
        },
      ],
      BoundingBox: {
        Height: 0.7783375382423401,
        Left: 0.062234796583652496,
        Top: 0.20151133835315704,
        Width: 0.7751060724258423,
      },
      Confidence: 96.86412811279297,
      Id: 0,
    },
  ],
  ProtectiveEquipmentModelVersion: "1.0",
};

describe("ppe Mapper", () => {
  test("maps with body parts and confidence", () => {
    const mapped = testData.Persons.map(ppeMapper);
    expect(mapped).toMatchSnapshot();
  });
});
