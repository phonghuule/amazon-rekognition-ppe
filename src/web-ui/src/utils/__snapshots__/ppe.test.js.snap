// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ppe Mapper maps with body parts and confidence 1`] = `
Array [
  Object {
    "boundingBox": Object {
      "Height": 0.7783375382423401,
      "Left": 0.062234796583652496,
      "Top": 0.20151133835315704,
      "Width": 0.7751060724258423,
    },
    "id": 0,
    "results": Array [
      Object {
        "bodyPart": "face",
        "boundingBox": Object {
          "Height": 0.3095385730266571,
          "Left": 0.2645637094974518,
          "Top": 0.29297786951065063,
          "Width": 0.1740061491727829,
        },
        "confidence": 99.9,
        "coversBodyPart": true,
        "coversBodyPartConfidence": 79.8,
        "type": "Mask",
      },
      Object {
        "bodyPart": "face",
        "boundingBox": Object {
          "Height": 0.3095385730266571,
          "Left": 0.2645637094974518,
          "Top": 0.29297786951065063,
          "Width": 0.1740061491727829,
        },
        "confidence": 99.9,
        "coversBodyPart": true,
        "coversBodyPartConfidence": 79.8,
        "type": "Foo",
      },
      Object {
        "bodyPart": "head",
        "boundingBox": Object {
          "Height": 0.37994688749313354,
          "Left": 0.342035174369812,
          "Top": 0.16647937893867493,
          "Width": 0.2135859727859497,
        },
        "confidence": 99.9,
        "coversBodyPart": true,
        "coversBodyPartConfidence": 84.2,
        "type": "Helmet",
      },
      Object {
        "bodyPart": "left hand",
        "boundingBox": Object {
          "Height": 0.33101364970207214,
          "Left": 0.5892353653907776,
          "Top": 0.23174917697906494,
          "Width": 0.18587328493595123,
        },
        "confidence": 99.7,
        "coversBodyPart": true,
        "coversBodyPartConfidence": 100,
        "type": "Glove",
      },
      Object {
        "bodyPart": "right hand",
        "boundingBox": Object {
          "Height": 0.32489436864852905,
          "Left": 0.20085042715072632,
          "Top": 0.2759649455547333,
          "Width": 0.18243715167045593,
        },
        "confidence": 99.8,
        "coversBodyPart": true,
        "coversBodyPartConfidence": 100,
        "type": "Glove",
      },
    ],
  },
]
`;
